### ISSUE TITLE
Invidious: startup fails with local database due to auth failures

### ISSUE THREAD

**Issue by isker** [NixOS/nixpkgs#157361]
### Describe the bug
With default configuration of [Invidious](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-apps/invidious.nix), startup fails due to auth problems setting up the postgres database for the service.

Full invidious.service logs:
```
Jan 30 01:06:01 systemd[1]: Started Invidious (An alternative YouTube front-end).
Jan 30 01:06:01 invidious-start[142062]: Unhandled exception:  (DB::ConnectionRefused)
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/x5fc02s9nq1qxrr2fgzpgqq1vckzb47i-invidious-unstable-2021-11-08/bin/invidious in '??'
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/x5fc02s9nq1qxrr2fgzpgqq1vckzb47i-invidious-unstable-2021-11-08/bin/invidious in '??'
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/x5fc02s9nq1qxrr2fgzpgqq1vckzb47i-invidious-unstable-2021-11-08/bin/invidious in '??'
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/x5fc02s9nq1qxrr2fgzpgqq1vckzb47i-invidious-unstable-2021-11-08/bin/invidious in '??'
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/x5fc02s9nq1qxrr2fgzpgqq1vckzb47i-invidious-unstable-2021-11-08/bin/invidious in '__crystal_main'
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/x5fc02s9nq1qxrr2fgzpgqq1vckzb47i-invidious-unstable-2021-11-08/bin/invidious in 'main'
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/41dj1v3qz9a5kjncpkxhmq50yg9r24dn-glibc-2.33-62/lib/libc.so.6 in '__libc_start_main'
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/x5fc02s9nq1qxrr2fgzpgqq1vckzb47i-invidious-unstable-2021-11-08/bin/invidious in '_start'
Jan 30 01:06:01 invidious-start[142062]:   from ???
Jan 30 01:06:01 invidious-start[142062]: Caused by: Peer authentication failed for user "kemal" (PQ::PQError)
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/x5fc02s9nq1qxrr2fgzpgqq1vckzb47i-invidious-unstable-2021-11-08/bin/invidious in '??'
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/x5fc02s9nq1qxrr2fgzpgqq1vckzb47i-invidious-unstable-2021-11-08/bin/invidious in '??'
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/x5fc02s9nq1qxrr2fgzpgqq1vckzb47i-invidious-unstable-2021-11-08/bin/invidious in '??'
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/x5fc02s9nq1qxrr2fgzpgqq1vckzb47i-invidious-unstable-2021-11-08/bin/invidious in '??'
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/x5fc02s9nq1qxrr2fgzpgqq1vckzb47i-invidious-unstable-2021-11-08/bin/invidious in '??'
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/x5fc02s9nq1qxrr2fgzpgqq1vckzb47i-invidious-unstable-2021-11-08/bin/invidious in '??'
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/x5fc02s9nq1qxrr2fgzpgqq1vckzb47i-invidious-unstable-2021-11-08/bin/invidious in '??'
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/x5fc02s9nq1qxrr2fgzpgqq1vckzb47i-invidious-unstable-2021-11-08/bin/invidious in '??'
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/x5fc02s9nq1qxrr2fgzpgqq1vckzb47i-invidious-unstable-2021-11-08/bin/invidious in '__crystal_main'
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/x5fc02s9nq1qxrr2fgzpgqq1vckzb47i-invidious-unstable-2021-11-08/bin/invidious in 'main'
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/41dj1v3qz9a5kjncpkxhmq50yg9r24dn-glibc-2.33-62/lib/libc.so.6 in '__libc_start_main'
Jan 30 01:06:01 invidious-start[142062]:   from /nix/store/x5fc02s9nq1qxrr2fgzpgqq1vckzb47i-invidious-unstable-2021-11-08/bin/invidious in '_start'
Jan 30 01:06:01 invidious-start[142062]:   from ???
Jan 30 01:06:01 systemd[1]: invidious.service: Main process exited, code=exited, status=1/FAILURE
Jan 30 01:06:01 systemd[1]: invidious.service: Failed with result 'exit-code'.
```
Relevant part of postgresql.service logs:
```
Jan 30 01:06:01 postgres[142078]: [142078] LOG:  could not look up local user ID 65186: user does not exist
Jan 30 01:06:01 postgres[142078]: [142078] FATAL:  Peer authentication failed for user "kemal"
Jan 30 01:06:01 postgres[142078]: [142078] DETAIL:  Connection matched pg_hba.conf line 1: "local invidious kemal peer map=invidious"
```

I am not sure what this exactly indicates.  Perhaps related to the systemd `DynamicUser` or `PrivateUsers` settings?

### Steps To Reproduce
`services.invidious.enable = true;` seems to be sufficient.

### Expected behavior
Invidious can authenticate with the automatically created local postgres database.

### Notify maintainers

@sbruder I'm curious how you're configuring this service - are you not using a local database?

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.93, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.16`
 - channels(root): `"nixos-21.11.335526.6c4b9f1a2fd"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by sbruder** [comment]
I just tried it on a fresh system and could reproduce it. 

However, the dynamic user doesn’t seem to be a problem in the NixOS test, which works successfully (when intentionally making it fail it just hangs as port 3000 will never be open).

You can find my configuration here: https://git.sbruder.de/simon/nixos-config/src/branch/master/machines/yuzuru/services/invidious/default.nix Though I just discovered that a real (not dynamic) `invidious` user exists on that machine for some reason, which could be the reason for why it works on it.

**Comment by sbruder** [comment]
I opened a PR that switches from using a dynamic user to a system user which worked for me. If you can, please test if that fix also works for you.

**Comment by isker** [comment]
@sbruder Thanks for the quick fix.  I'd love to test, but I'm not sure how 😅.  I only started using nixos yesterday.  I'm not sure what the proper way to use your branch's definition of the service is. (I'm not sure that I should switch channels to your branch?  That'd change everything, right?  I'm using 21.11 for what it's worth...).

If you don't mind walking me through it...

**Comment by sbruder** [comment]
You are right, using the PR’s branch as your main channel is not the best option, because it could include unrelated changes that might permanently change state. Using a module from a different nixpkgs version can instead be done by using the `disabledModules` option (see the [NixOS manual](https://nixos.org/manual/nixos/stable/index.html#sec-replace-modules) if you want to know more about it) and importing the specific module from the nixpkgs fork.

In this case the following should work when added to the NixOS configuration (assuming you don’t use flakes):

```nix
{
  imports = [
    "${builtins.fetchTarball "https://github.com/sbruder/nixpkgs/archive/refs/heads/fix-invidious-dynamicuser.tar.gz"}/nixos/modules/services/web-apps/invidious.nix"
  ];
  disabledModules = [ "services/web-apps/invidious.nix" ];
}
```

Please note that this is not a permanent solution, because the URL is not stable (i.e. its contents can change).

**Comment by isker** [comment]
Thanks for the explanation.  It works!

If you don't mind explaining more, is this a thing that'd be acceptable to backport, or would I instead use unstable for this module, like in the manual section you linked?

**Comment by isker** [comment]
Ah, well, I see you added a backport label to the PR.

**Comment by sbruder** [comment]
I think the PR is acceptable to be backported, because it currently doesn’t work in 21.11. Once the PR is merged and the backport PR is created, you can track the progress for it to become available in 21.11 here: https://nixpk.gs/pr-tracker.html

**Comment by tilpner** [comment]
I've encountered this issue elsewhere, and it turned out nscd.service wasn't running (possibly addressed by #154928). If that also fixes your issue, then DynamicUser should probably be kept.

**Comment by isker** [comment]
@tilpner you're right.  I manually started `nscd.service` and everything works with the stable version of this package.

Welp.

**Comment by tilpner** [comment]
A different fix (https://github.com/NixOS/nixpkgs/pull/154620) has already been merged into release-21.11, but the channel hasn't advanced far enough yet.

**Comment by sbruder** [comment]
I can confirm that https://github.com/NixOS/nixpkgs/pull/154620, which now is in nixos-21.11, fixes this issue. Thank you @tilpner for your help!

**Comment by isker** [comment]
Thanks to you both!
